File: apps/src/lib/api/transport-web-commands/proxy-profiles.ts
import type { WebCommandDescriptor } from "./shared";

export function createProxyProfilesWebCommands(): Record<string, WebCommandDescriptor> {
  return {
    service_system_proxy_list: { rpcMethod: "system/proxy/list" },
    service_system_proxy_test_presets: { rpcMethod: "system/proxy/test-presets" },
    service_system_proxy_create: { rpcMethod: "system/proxy/create" },
    service_system_proxy_update: { rpcMethod: "system/proxy/update" },
    service_system_proxy_delete: { rpcMethod: "system/proxy/delete" },
    service_system_proxy_test_latency: { rpcMethod: "system/proxy/test-latency" },
    service_system_proxy_speed_test: { rpcMethod: "system/proxy/speed-test" },
    service_system_proxy_cloudflare_speed_test: { rpcMethod: "system/proxy/cloudflare-speed-test" },
    service_system_proxy_test_job: { rpcMethod: "system/proxy/test-job" },
    service_system_proxy_cancel_test: { rpcMethod: "system/proxy/cancel-test" },
    service_system_proxy_speed_test_history: { rpcMethod: "system/proxy/speed-test-history" },
    service_system_proxy_latency_test_history: { rpcMethod: "system/proxy/latency-test-history" },
    service_system_proxy_diagnostics_history: { rpcMethod: "system/proxy/diagnostics-history" },
    service_system_proxy_import_batch: { rpcMethod: "system/proxy/import-batch" },
    service_system_proxy_pool_list: { rpcMethod: "system/proxy-pool/list" },
    service_system_proxy_pool_create: { rpcMethod: "system/proxy-pool/create" },
    service_system_proxy_pool_update: { rpcMethod: "system/proxy-pool/update" },
    service_system_proxy_pool_delete: { rpcMethod: "system/proxy-pool/delete" },
    service_system_proxy_pool_members_add: { rpcMethod: "system/proxy-pool/members/add" },
    service_system_proxy_pool_members_remove: { rpcMethod: "system/proxy-pool/members/remove" },
    service_system_proxy_pool_members_set_enabled: { rpcMethod: "system/proxy-pool/members/set-enabled" },
    service_system_proxy_pool_health_check: { rpcMethod: "system/proxy-pool/health-check" },
    service_system_proxy_pool_switch_logs: { rpcMethod: "system/proxy-pool/switch-logs" },
  };
}

File: apps/src-tauri/src/commands/settings/proxy_profiles.rs
use crate::commands::shared::rpc_call_in_background;

#[tauri::command]
pub async fn service_system_proxy_list(
    addr: Option<String>,
) -> Result<serde_json::Value, String> {
    rpc_call_in_background("system/proxy/list", addr, None).await
}

#[tauri::command]
pub async fn service_system_proxy_test_presets(
    addr: Option<String>,
) -> Result<serde_json::Value, String> {
    rpc_call_in_background("system/proxy/test-presets", addr, None).await
}

#[tauri::command]
pub async fn service_system_proxy_create(
    addr: Option<String>,
    name: String,
    proxy_url: String,
    enabled: Option<bool>,
    tags_json: Option<String>,
    notes: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "name": name,
        "proxyUrl": proxy_url,
        "enabled": enabled,
        "tagsJson": tags_json,
        "notes": notes,
    });
    rpc_call_in_background("system/proxy/create", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_update(
    addr: Option<String>,
    id: String,
    name: Option<String>,
    proxy_url: Option<String>,
    enabled: Option<bool>,
    tags_json: Option<String>,
    notes: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "name": name,
        "proxyUrl": proxy_url,
        "enabled": enabled,
        "tagsJson": tags_json,
        "notes": notes,
    });
    rpc_call_in_background("system/proxy/update", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_delete(
    addr: Option<String>,
    id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "id": id });
    rpc_call_in_background("system/proxy/delete", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_test(
    addr: Option<String>,
    id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({ "id": id });
    rpc_call_in_background("system/proxy/test", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_test_latency(
    addr: Option<String>,
    id: String,
    preset_id: Option<String>,
    custom_url: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "presetId": preset_id,
        "customUrl": custom_url,
    });
    rpc_call_in_background("system/proxy/test-latency", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_speed_test(
    addr: Option<String>,
    id: String,
    provider_id: Option<String>,
    file_size_id: Option<String>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "providerId": provider_id,
        "fileSizeId": file_size_id,
    });
    rpc_call_in_background("system/proxy/speed-test", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_cloudflare_speed_test(
    addr: Option<String>,
    id: String,
    config: Option<serde_json::Value>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "config": config,
    });
    rpc_call_in_background("system/proxy/cloudflare-speed-test", addr, Some(params)).await
}


#[tauri::command]
pub async fn service_system_proxy_test_job(
    addr: Option<String>,
    job_id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "jobId": job_id,
    });
    rpc_call_in_background("system/proxy/test-job", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_cancel_test(
    addr: Option<String>,
    job_id: String,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "jobId": job_id,
    });
    rpc_call_in_background("system/proxy/cancel-test", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_speed_test_history(
    addr: Option<String>,
    id: String,
    limit: Option<usize>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "limit": limit,
    });
    rpc_call_in_background("system/proxy/speed-test-history", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_latency_test_history(
    addr: Option<String>,
    id: String,
    limit: Option<usize>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "limit": limit,
    });
    rpc_call_in_background("system/proxy/latency-test-history", addr, Some(params)).await
}

#[tauri::command]
pub async fn service_system_proxy_diagnostics_history(
    addr: Option<String>,
    id: String,
    limit: Option<usize>,
) -> Result<serde_json::Value, String> {
    let params = serde_json::json!({
        "id": id,
        "limit": limit,
    });
    rpc_call_in_background("system/proxy/diagnostics-history", addr, Some(params)).await
}
